compiler/passes/errorHandling.cpp

```diff
@@ -271,57 +271,42 @@ void ErrorHandlingVisitor::lowerCatches(const TryInfo& info) {
     SET_LINENO(c);
 
     CatchStmt* catchStmt = toCatchStmt(c);
-    BlockStmt* catchBody = catchStmt->body();
-    DefExpr*   catchDef  = catchStmt->expr();
 
-    VarSymbol* toDelete = errorVar;
+    BlockStmt* catchBody = catchStmt->body();
 
-    catchBody->remove();
+    // PRIM_CURRENT_ERROR should be already replaced by info.errorVar
 
-    // catchall
-    if (catchDef == NULL) {
+    // named or unnamed catchall
+    if (catchStmt->isCatchall()) {
       hasCatchAll = true;
-      currHandler->insertAtTail(catchBody);
+      currHandler->insertAtTail(catchBody->remove());
     } else {
-      VarSymbol* errSym  = toVarSymbol(catchDef->sym);
-      Type*      errType = errSym->type;
-
-      toDelete = errSym;
-      catchDef->remove();
-      currHandler->insertAtTail(catchDef);
+      // Find the else body in the last CondStmt in the catch block body.
+      // This should always exist after CatchStmt::cleanup
+      // for non-catchall errors.
+
+      CondStmt* finalCond = NULL;
+      for_alist_backward(node, catchBody->body) {
+        if (CondStmt* cond = toCondStmt(node)) {
+          finalCond = cond;
+          break;
+        }
+      }
+      INT_ASSERT(finalCond != NULL && finalCond->elseStmt != NULL);
 
-      // named catchall
-      if (errType == dtError) {
-        hasCatchAll = true;
-        currHandler->insertAtTail(new CallExpr(PRIM_MOVE, errSym, errorVar));
-        currHandler->insertAtTail(errorCond(errSym, catchBody));
+      BlockStmt* nextHandler = finalCond->elseStmt;
 
-      // specified catch
-      } else {
-        CallExpr*  castError   = new CallExpr(PRIM_DYNAMIC_CAST,
-                                              new SymExpr(errType->symbol),
-                                              errorVar);
-        BlockStmt* nextHandler = new BlockStmt();
+      // Clear everything in the elseStmt
+      // (sometimes has PRIM_RT_ERROR to satisfy isDefinedAllPaths)
+      for_alist(stmt, nextHandler->body)
+        stmt->remove();
```

Comment:
two reasons, but I don't think it matters much either way.

1. I didn't want to worry about when I compute nextHandler - literally replacing it here would cause problems because nextHandler would be out of the AST. But of course we could shuffle the order around.... but then it's more complex / fiddly code.
2. I generally don't like to introduce new AST nodes if I can avoid it (saves memory, saves time in tracking AST ids backwards through passes).